src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,303 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Analyzer.Utilities;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeQuality.Analyzers.ApiDesignGuidelines
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic), Shared]
+    public sealed class StringBuilderAppendShouldNotTakeSubstringFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray
+            .Create(
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdOneParameterId,
+                StringBuilderAppendShouldNotTakeSubstring.RuleIdTwoParameterId);
+
+        public sealed override FixAllProvider GetFixAllProvider()
+        {
+            return StringBuilderAppendShouldNotTakeSubstringFixAllProvider.Instance;
+        }
+
+        private class StringBuilderAppendShouldNotTakeSubstringFixAllProvider : FixAllProvider
```

Comment:
>StringBuilderAppendShouldNotTakeSubstringFixAllProvider [](start = 22, length = 55)

I am not sure why you need a special FixAllProvider for this code fix. Please just switch to WellKnownBatchFixer that is used across this repo.

Reply:
I'd be very very happy to do that, but the BatchFixer doesn't work here.
Using that, the following tests fail:

1. FixesAllInChainedAppends()
2. FixesAllInChainedAppendsBasic
3. FixesAllOnExampleFromTicket (!!)
4. FixesAllOnExampleFromTicketBasic (!!)

But IMHO especially chaining Appends on the StringBuilder is a quite common use case, so I'd like to solve that case as well - as the current custom fixer does.

Reply:
> especially chaining Appends on the StringBuilder is a quite common use case, so I'd like to solve that case as well - as the current custom fixer does.

Thanks for the detailed investigation and unit tests here. However, the approach chosen by this repo right from the start has been to use batch fixer, which works fine for all cases where there are no conflicting edits. This decision has been driven by the fact that the value add from handling corner cases where there are conflicting fixes is not worth the huge complexity of writing custom FixAllProvider + unit tests that handle these, and if we do get real customer requests where this is turning out to be too much of a hindrance, then we will go and right a custom FixAllProvider. Till date, we haven't had a single report of batch fixer not serving end user's purposes, so that sort of justifies that claim. I'd prefer to stick to this approach and come back to add this custom FixAllProvider if we have customer requests. Can you please modify/skip your tests accordingly and switch to batch fixer?